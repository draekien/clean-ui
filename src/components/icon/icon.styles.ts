import { ThemeUIStyleObject } from 'theme-ui';
import { Size } from '../../types/layouts';
import { colors } from '../theme/theme.colors';

export interface IconCssProps {
  size: Size;
  color?: keyof typeof colors;
}

export const iconCss = ({ size, color }: IconCssProps): ThemeUIStyleObject => {
  const css: any = {
    verticalAlign: 'middle',
    color: color || 'inherit',
  };

  switch (size) {
    case 'small':
      css.fontSize = '1rem !important';
      break;
    case 'medium':
      css.fontSize = '1.5rem !important';
      break;
    case 'large':
      css.fontSize = '2rem !important';
      break;
  }

  return css;
};
