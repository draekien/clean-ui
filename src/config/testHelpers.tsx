import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
