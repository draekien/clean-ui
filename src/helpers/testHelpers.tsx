import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
