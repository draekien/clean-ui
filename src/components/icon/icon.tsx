/** @jsxImportSource theme-ui */
import React from 'react';
import { Size } from '../../types/layouts';
import { colors } from '../theme/theme.colors';
import * as styles from './icon.styles';

export type IconVariant = 'filled' | 'outlined' | 'round' | 'two-tone' | 'sharp';

export interface IconProps extends React.HTMLAttributes<HTMLElement> {
  /** the name of the material-icon */
  name: string;
  /** the material icon variant
   * @default 'filled'
   */
  variant?: IconVariant;
  /** the icon size
   * @default 'medium'
   */
  size?: Size;
  /** the color of the icon */
  color?: keyof typeof colors;
}

const getIconClassName = (variant?: IconVariant): string => {
  switch (variant) {
    case 'outlined':
    case 'round':
    case 'sharp':
    case 'two-tone':
      return `material-icons-${variant}`;
    default:
      return 'material-icons';
  }
};

export const Icon: React.FC<IconProps> = (props: IconProps) => {
  const { name, color, size = 'medium', variant = 'filled', className, ...rest } = props;

  return (
    <i
      sx={styles.iconCss({ size, color })}
      className={`${getIconClassName(variant)} ${className || ''}`}
      {...rest}>
      {name}
    </i>
  );
};
