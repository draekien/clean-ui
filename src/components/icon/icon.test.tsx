import React from 'react';
import { themedRender } from '../../helpers/testHelpers';
import { Icon } from './icon';

describe('Icon Component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(<Icon name="accessibility" />);

    expect(container).toMatchSnapshot();
  });
});
