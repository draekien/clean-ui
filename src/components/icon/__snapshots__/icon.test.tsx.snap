// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Component Should match snapshot 1`] = `
<div>
  <i
    class="material-icons  css-zif8ik-Icon"
  >
    accessibility
  </i>
</div>
`;
